utils/riff/recommendations/rec_base.js

```diff
@@ -0,0 +1,262 @@
+    }
+
+    /** readonly isComplete property
+     *  true if the recommendation has been performed by the user, false if they
+     *  haven't performed the recommendation (or this recommendation isn't
+     *  initialized).
+     *  @type {bool}
+     */
+    get isComplete() {
+        return this._isComplete;
+    }
+
+    /* **************************************************************************
+     * initialize                                                          */ /**
+     *
+     * Initialize this recommendation instance. Because a recommendation may
+     * depend on asynchronous results before it has been initialized, this
+     * method returns a promise also accessible via the isInitialized property.
+     *
+     * This base implementation needs no initialization it should be overridden
+     * in derived classes which do require initialization.
+     *
+     * @param {function(): Object} getState - function which returns the redux state
+     *
+     * @returns {Promise<bool>} true if this recommendation has been initialized.
+     */
+    initialize(getState) {
+        if (this._isInitialized === null) {
+            this._isInitialized = this.doInitialization(getState);
+        }
+
+        return this._isInitialized;
+    }
+
+    /* **************************************************************************
+     * doInitialization                                                    */ /**
+     *
+     * Does all initialization (possibly asynchronous) of this recommendation
+     * instance.
+     *
+     * This base implementation calls and awaits the protected function checkIsComplete
+     * to set the isComplete property. If more things need to be initialized this
+     * method should be overridden in the derived class to do it.
+     *
+     * DevNote: This method exists so that the initialize method can set the
+     * _isInitialized instance property to the same promise that it returns.
+     * It should only be called by the base class initialize method.
+     * @protected
+     *
+     * @param {function(): Object} getState - function which returns the redux state
+     *
+     * @returns {Promise<bool>} true if this recommendation has been initialized.
+     */
+    async doInitialization(getState) {
+        await this.checkIsComplete(getState);
+        return true;
+    }
+
+    /* **************************************************************************
+     * checkIsComplete                                                     */ /**
+     *
+     * Tests if the current user has completed this recommendation, and sets
+     * the {@link RecommendationBase#isComplete} property to reflect what is
+     * determined.
+     * This base implementation always sets the value to false.
+     * This method should NOT be called if the recommendation is not
+     * displayable see {@link shouldDisplay}.
+     * Derived recommendations should override this implementation AND not
+     * call it.
```

Comment:
I'm a little confused, but I think it is twofold. 
1. If `shouldDisplay()` returns false, then don't call `checkIsComplete()`.
2. I'm not too sure what the derived recommendations means and what overriding means.

Reply:
@brecriffs I'm confused about lines 156 & 157. A derived recommendation is one that inherits from the RecommendationBase class (i.e. it is a sub-type of the RecommendationBase). The comments say that derived recommendation classes should override this implementation. That means they should replace this definition of the function with their own, different definition. What's not clear to me is that it says derived classes should override AND not call. I think this means that they should make sure that the overriding definitions _also_ do not get called unless the recommendation is displayable, but it could also be read as saying the overriding definition should _never_ be called, so I wanted to double check the intention.

Reply:
Really I should have just labeled this method as `@abstract` and not provided an implementation in the base class at all.

What I was trying to express is that this base implementation does nothing useful and therefore the override method shouldn't bother to call it.

For example the base `shouldDisplay` method handles the regular case of determining if the recommendation should be displayed based on its `startWeek` and `numWeeksToDisplay`. There are derived classes that override `shouldDisplay` to check additional constraints specific to themselves. They still call the base implementation to perform the check of `startWeek` and `numWeeksToDisplay`.

Reply:
Hmm interesting. Yeah thinking back to college, this sounds like an abstract method. Everything here is very OO, and I need to brush up a bit on those skills!